### PR TITLE
feat(rule): aria-roledescription

### PR DESCRIPTION
**Pull request by straker**
Aria-roledescription required [too many nuances](https://github.com/dequelabs/axe-core/issues/1216#issuecomment-454104190) that couldn't adequately be captured in just an `unsupported` object:

- required matching a list of supported roles
- matcher objects don't allow matching of elements with implicit roles
- required failing on any element without a role

As such, it was easier to turn it into its own rule that could capture these nuances. This is the same problem we'll encounter when trying to [support `role="text"`](https://github.com/dequelabs/axe-core/issues/1597) as we'd want to make sure that it isn't used on an element with interactive children (something else that can't be captured in an `unsupported` object).

Closes issue: #1517

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Code is reviewed for security

### CODE REVIEW

File: lib/rules/aria-roledescription.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+	"id": "aria-roledescription",
+	"selector": "[aria-roledescription]",
+	"excludeHidden": false,
+	"tags": ["cat.aria", "best-practice"],
```
Comment (dylanb) [inline lib/rules/aria-roledescription.json]: Need to engage with the Jedi Council to determine whether this should be categorized under WCAG 4.1.2

My initial reaction is that it should because if used incorrectly, it means that the role is not being described correctly.

File: lib/checks/aria/aria-roledescription.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+const role = axe.commons.aria.getRole(node);
+const supportedRoles = [
```
Comment (dylanb) [inline lib/checks/aria/aria-roledescription.js]: Should we not make these configurable options? This way it can be changed in custom rules to

1. Support different AT environments (e.g. internal employee application with a controlled AT environment), and
1. Be changed easily in custom rules if support changes without having to wait for an axe-core release

Comment (straker) [reply lib/checks/aria/aria-roledescription.js]: Good point

File: lib/rules/aria-roledescription.json [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+	"metadata": {
+		"description": "Ensures aria-roledescription attribute is used correctly",
+		"help": "Aria-roledescription attribute must be used on elements with an implicit or explicit role"
```
Comment (WilcoFiers) [inline lib/rules/aria-roledescription.json]: The help text is supposed to be the shorter text. I also don't think this conveys quite enough information. Suggestion:

> description: Ensure aria-roledescription is only used on elements with an implicit or explicit role
> help: Use aria-roledescription on elements with a semantic role

File: lib/checks/aria/aria-roledescription.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+			"combobox",
+			"menuitemcheckbox",
+			"menuitemradio"
+		]
+	},
+	"metadata": {
+		"impact": "serious",
+		"messages": {
+			"pass": "Used on a known supported role",
+			"incomplete": "Check that the aria-roledescription is announced",
```
Comment (WilcoFiers) [inline lib/checks/aria/aria-roledescription.json]: can you add " by supported screen readers"?

File: lib/checks/aria/aria-roledescription.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+			"radio",
+			"combobox",
+			"menuitemcheckbox",
+			"menuitemradio"
+		]
+	},
+	"metadata": {
+		"impact": "serious",
+		"messages": {
+			"pass": "Used on a known supported role",
```
Comment (WilcoFiers) [inline lib/checks/aria/aria-roledescription.json]: Use full sentences please.

File: lib/checks/aria/aria-roledescription.json [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+			"incomplete": "Check that the aria-roledescription is announced",
+			"fail": "Used on an element without an implicit or explicit role"
```
Comment (WilcoFiers) [inline lib/checks/aria/aria-roledescription.json]: I suggest "Give the element a role that supports aria-roledescription"

File: lib/checks/aria/aria-roledescription.js [hunk 7]
```diff
@@ -0,0 +1,14 @@
+options = options || {};
+
+const role = axe.commons.aria.getRole(node);
+const supportedRoles = options.supportedRoles || [];
+
+if (supportedRoles.includes(role)) {
+	return true;
+}
+
+if (role) {
```
Comment (WilcoFiers) [inline lib/checks/aria/aria-roledescription.js]: Should we check if the role is `presentation|none` and fail those?

File: lib/rules/aria-roledescription.json (hunk 1 above)
Comment (WilcoFiers) [reply lib/rules/aria-roledescription.json]: I'd say keep the tags consistent with what it was before we split out: `"cat.aria", "wcag2a", "wcag412"`

File: test/integration/rules/aria-roledescription/aria-roledescription.html [hunk 8]
```diff
@@ -0,0 +1,9 @@
+<img aria-roledescription="my img" src="foo.png" id="pass1" />
+<button aria-roledescription="my button" id="pass2">button</button>
+<div role="combobox" aria-roledescription="my combobox" id="pass3"></div>
```
Comment (WilcoFiers) [inline test/integration/rules/aria-roledescription/aria-roledescription.html]: Can you put in cases for each of the check options?

File: lib/rules/aria-roledescription.json [hunk 9]
```diff
@@ -0,0 +1,13 @@
+{
+	"id": "aria-roledescription",
+	"selector": "[aria-roledescription]",
+	"excludeHidden": false,
```
Comment (WilcoFiers) [inline lib/rules/aria-roledescription.json]: This is different from `aria-allowed-attr`, why did you change this? I don't think this is likely to cause false positives, but probably not any less than `aria-allowed-attr`. IMO it's probably safer to not do this and rely on people testing axe in multiple page states instead.

File: lib/checks/aria/aria-roledescription.js [hunk 10]
```diff
@@ -0,0 +1,14 @@
+options = options || {};
```
Comment (jkodu) [inline lib/checks/aria/aria-roledescription.js]: nitpick, destructuring will be easier to read & will remove line 4.

```
const { supportedRoles = [] } = options
```

File: lib/checks/aria/aria-roledescription.js [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+const role = axe.commons.aria.getRole(node);
+const supportedRoles = options.supportedRoles || [];
+
+if (supportedRoles.includes(role)) {
+	return true;
+}
+
+if (role && role !== 'presentation' && role !== 'none') {
```
Comment (jkodu) [inline lib/checks/aria/aria-roledescription.js]: nitpick, exit early again easier to read.

```
if(!role) {
  return false
}

if(!['presentation', 'none'].includes(role)) {
 return undefined
}
```
